File: node/src/rate-limit.mjs
export class RateLimiter {
  constructor({ limit, windowMs, clock = Date.now }) {
    this.limit = limit;
    this.windowMs = windowMs;
    this.clock = clock;
    this.entries = new Map();
  }

  take(key) {
    const now = this.clock();
    const current = this.entries.get(key);
    if (!current || current.resetAt <= now) {
      this.entries.set(key, { count: 1, resetAt: now + this.windowMs });
      this.#prune(now);
      return true;
    }
    if (current.count >= this.limit) return false;
    current.count += 1;
    return true;
  }

  #prune(now) {
    if (this.entries.size < 1_000) return;
    for (const [key, entry] of this.entries) {
      if (entry.resetAt <= now) this.entries.delete(key);
    }
  }
}

File: node/internal/ratelimit/ratelimit.go
package ratelimit

import (
	"sync"
	"time"
)

type entry struct {
	count int
	reset time.Time
}
type Limiter struct {
	mu      sync.Mutex
	limit   int
	window  time.Duration
	now     func() time.Time
	entries map[string]entry
}

func New(limit int, window time.Duration, now func() time.Time) *Limiter {
	return &Limiter{limit: limit, window: window, now: now, entries: map[string]entry{}}
}
func (l *Limiter) Take(key string) bool {
	l.mu.Lock()
	defer l.mu.Unlock()
	now := l.now()
	e, ok := l.entries[key]
	if !ok || !now.Before(e.reset) {
		l.entries[key] = entry{count: 1, reset: now.Add(l.window)}
		if len(l.entries) >= 1000 {
			for k, v := range l.entries {
				if !now.Before(v.reset) {
					delete(l.entries, k)
				}
			}
		}
		return true
	}
	if e.count >= l.limit {
		return false
	}
	e.count++
	l.entries[key] = e
	return true
}
